src/charm.py

```diff
@@ -9,264 +9,93 @@
-        try:
-            return self._tempo_role
-        except ValueError:
-            return None
-
-    def _update_tls_certificates(self) -> bool:
-        # caller is responsible for guarding for container connectivity
-        if privkey := self.tempo_cluster.get_privkey():
-            ca_cert, server_cert = self.tempo_cluster.get_ca_and_server_certs()
-            self.tempo.update_certs(privkey, ca_cert, server_cert)
-            if ca_cert:
-                # update cacert in charm container too, for self-instrumentation
-                Path(TEMPO_CA_FILE).write_text(ca_cert)
-                subprocess.run(["update-ca-certificates", "--fresh"])
-            return True
-
-        else:
-            self.tempo.clear_certs()
-
-            # clear from charm container too
-            Path(TEMPO_CA_FILE).unlink(missing_ok=True)
-            subprocess.run(["update-ca-certificates", "--fresh"])
-            return True
-
-    def pre_update_checklist(
-        self,
-    ) -> List[str]:
-        """Check if the charm as a whole is ready, and if not, return a list of failed checks."""
-        failures = []
-        failures.extend(self.tempo.pre_update_checklist())
-
-        if not self.tempo_cluster.get_tempo_config():
-            failures.append("tempo_cluster config not ready")
-
-        return failures
-
     def _on_collect_status(self, e: CollectStatusEvent):
-        if not self._container.can_connect():
-            e.add_status(WaitingStatus("Waiting for `tempo` container"))
-        if not self.model.get_relation("tempo-cluster"):
-            e.add_status(
-                BlockedStatus("Missing tempo-cluster relation to a tempo-coordinator charm")
-            )
-        elif not self.tempo_cluster.relation:
-            e.add_status(WaitingStatus("Tempo-Cluster relation not ready"))
-        if not self.tempo_cluster.get_tempo_config():
-            e.add_status(WaitingStatus("Waiting for coordinator to publish a tempo config"))
-
-        if role := self.tempo_role:
-            e.add_status(
-                ActiveStatus(
-                    "(all roles) ready." if role is TempoRole.all else f"{role.name} ready."
-                )
-            )
-        else:
-            logger.error(
-                f"`role` config value {self.config.get('role')!r} invalid: should "
-                f"be one of {list(TempoRole)}."
-            )
+        # add Tempo worker custom blocking conditions
+        if self.worker.roles and self.worker.roles[0] not in self._valid_roles:
             e.add_status(
                 BlockedStatus(f"Invalid `role` config value: {self.config.get('role')!r}.")
             )
 
     @property
     def tempo_endpoint(self) -> Optional[str]:
         """Tempo endpoint for charm tracing."""
-        return self.tempo_cluster.get_tracing_endpoint("otlp_http")
+        if endpoints := self.worker.cluster.get_tracing_receivers():
```

Comment:
please switch to using the [charm_tracing_config from the tracing lib](https://github.com/canonical/tempo-k8s-operator/blob/main/lib/charms/tempo_k8s/v2/tracing.py#L928).

Reply:
This is different. In the worker, we don't have a `TracingEndpointRequirer` anymore in the genetic worker. The tracing receivers are passed in relation data from the coordinator. That's why there is now a `get_tracing_receivers` exposed.

Reply:
ah! right. then for now it's ok, I'll contribute something to the Worker to abstract this logic